src/sql/src/plan/statement/dml.rs

```diff
@@ -160,25 +166,128 @@ pub fn plan_read_then_write(
+                    offset: None,
+                },
+            as_of: None,
+        } if ctes.is_empty() && order_by.is_empty() => match std::ops::Deref::deref(body) {
+            Select {
+                distinct: None,
+                projection,
+                from,
+                selection: None,
+                group_by,
+                having: None,
+                options,
+            } if from.is_empty()
+                && group_by.is_empty()
+                && options.is_empty()
+                && projection.len() == 1 =>
+            {
+                match projection.as_slice() {
+                    [SelectItem::Expr {
+                        expr:
+                            Expr::Function(Function {
+                                name,
+                                args: FunctionArgs::Args { args, order_by },
+                                filter: None,
+                                over: None,
+                                distinct: false,
+                            }),
+                        alias: None,
+                    }] if order_by.is_empty() => {
+                        fn eval_expr<'a>(
+                            ecx: &ExprContext,
+                            expr: &Expr<Aug>,
+                            ty: &ScalarType,
+                        ) -> Result<mz_repr::Row, PlanError> {
+                            let scalar = query::plan_expr(ecx, expr)?;
+                            let hir = scalar.cast_to(ecx, CastContext::Assignment, ty)?;
+                            let mir = hir.lower_uncorrelated()?;
+                            let temp_storage = mz_repr::RowArena::new();
+                            let datum = mir.eval(&[], &temp_storage)?;
+                            Ok(mz_repr::Row::pack_slice(&[datum]))
+                        }
+
+                        let pcx = plan::PlanContext::zero();
+                        let qcx = QueryContext::root(scx, QueryLifetime::OneShot(&pcx));
+                        let ecx = ExprContext {
+                            qcx: &qcx,
+                            name: "side effecting function planner",
+                            scope: &Scope::empty(),
+                            relation_type: &RelationType::empty(),
+                            allow_aggregates: false,
+                            // TODO: false here prevents parameters from working. But setting it to
+                            // true and using a param returns "unexpected ScalarExpr in uncorrelated
+                            // plan" at some point. Make params work.
+                            allow_subqueries: false,
+                            allow_windows: false,
+                        };
+                        match name.full_name_str().as_str() {
+                            "pg_catalog.pg_cancel_backend" if args.len() == 1 => {
+                                let connection_id = eval_expr(&ecx, &args[0], &ScalarType::Int32)?
```

Comment:
Does `eval_expr` return an error if the arg isn't an `Int32`?

Reply:
Yes, the call to `cast_to` will error if the cast is impossible!